SIMPLE LINEAR REGRESION

- In this lab we will use simple regression to predict car price.
- Objective:
    - to know if the dealer is offering the fair value for my trade-in.
    - To know if I put a fair value on my car.

DATASET

- We will use the dataset of Car price prediction hosted on IBM Cloud object.

IMPORT PACKAGES


In [4]:
import os #  for file operations
import numpy as np #  for numerical processing
import pandas as pd #   data frames
import matplotlib.pyplot as plt #  plotting library
import seaborn as sns #   statistical data visualization
from sklearn.model_selection import train_test_split # modeling
from sklearn.linear_model import LinearRegression # modeling
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score #modeling

In [5]:
Data_path = '/'.join(os.path.abspath('').split('/')[:-4])+'automobileEDA.csv'
df = pd.read_csv(Data_path)
df.head()

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

In [6]:
# get features
x = df.drop('price', axis=1) # independent features
y = df['price'] # dependent features

TRAIN TEST SPLIT

In [9]:
# random_state = shuffles from the fixed position

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1)

SIMPLE LINEAR REGRESSION

Predict car price using "highway.mpg"

In [12]:
# get simple linear independent variables
x_train_simple = x_train['highway-mpg'].values.reshape(-1, 1)
x_test_simple = x_test['highway-mpg'].values.reshape(-1, 1)

In [16]:
x_train_simple

array([[24],
       [27],
       [32],
       [22],
       [32],
       [43],
       [38],
       [33],
       [25],
       [31],
       [22],
       [41],
       [32],
       [23],
       [34],
       [46],
       [34],
       [30],
       [25],
       [32],
       [18],
       [37],
       [32],
       [29],
       [53],
       [25],
       [29],
       [22],
       [38],
       [25],
       [25],
       [23],
       [24],
       [24],
       [38],
       [30],
       [30],
       [28],
       [30],
       [37],
       [25],
       [47],
       [25],
       [24],
       [38],
       [28],
       [25],
       [32],
       [28],
       [34],
       [19],
       [29],
       [33],
       [29],
       [24],
       [36],
       [33],
       [23],
       [23],
       [31],
       [24],
       [38],
       [38],
       [37],
       [37],
       [26],
       [16],
       [30],
       [32],
       [25],
       [24],
       [27],
       [42],
       [30],
       [33],
       [25],
       [23],

In [14]:
x_test_simple

array([[34],
       [38],
       [32],
       [34],
       [30],
       [38],
       [24],
       [32],
       [22],
       [43],
       [34],
       [34],
       [30],
       [28],
       [23],
       [24],
       [54],
       [27],
       [37],
       [34],
       [33],
       [34],
       [20],
       [24],
       [42],
       [19],
       [30],
       [31],
       [32],
       [37],
       [28],
       [30],
       [28],
       [37],
       [24],
       [26],
       [30],
       [33],
       [22],
       [25],
       [38]], dtype=int64)

In [17]:
# Fit simple linear regression model
simple_lm =LinearRegression()
simple_lm.fit(x_train_simple, y_train)

LinearRegression()

In [18]:
# get prediction
y_hat_simple = simple_lm.predict(x_test_simple)
y_hat_simple[:5]


array([10530.19044542,  7150.17926639, 12220.19603494, 10530.19044542,
       13910.20162445])

In [19]:
# get intercept
simple_lm.intercept_

39260.28546717645

In [20]:
# get slope
simple_lm.coef_.item()

-845.0027947574558

ESTIMATED AMOUNT

    - Price = 39260.28546717645(simple_lm.intercept_) - 845.0027947574558(simple_lm.coef_.item()) * highway-mpg

Exercise
- Fit simple linear regression with horse_power ad independent variable to predict the price

Simple linear  regression model uing horsepower

In [21]:
# get simple linear independent variables
x_train_simple = x_train['horsepower'].values.reshape(-1, 1)
x_test_simple = x_test['horsepower'].values.reshape(-1, 1)

In [22]:
simple_lm = LinearRegression()
simple_lm.fit(x_train_simple, y_train)

LinearRegression()

In [23]:
y_hat_simple = simple_lm.predict(x_test_simple)
y_hat_simple[:5]

array([12166.10030237,  7173.74640389,  9928.14855478,  8550.94747933,
       15436.95285656])

In [24]:
simple_lm.intercept_

-4532.462737395461

In [25]:
simple_lm.coef_.item()

172.15013443062284

ESTIMATED AMOUNT

    - Price = -4532.462737395461(simple_lm.intercept_) + 172.15013443062284(simple_lm.coef_.item()) * horsepower

Exercise
- Fit simple linear regression with engine_size as independent variable to predict the price

Simple linear  regression model uing engine_size


In [26]:
x_train_simple = x_train['engine-size'].values.reshape(-1, 1)
x_test_simple = x_test['engine-size'].values.reshape(-1, 1)

In [27]:
simple_lm = LinearRegression()
simple_lm.fit(x_train_simple, y_train)

LinearRegression()

In [28]:
y_hat_simple = simple_lm.predict(x_test_simple)
y_hat_simple[:5]

array([12200.90004806,  7203.38742473, 12534.06755628,  7536.55493295,
       16532.07765495])

In [29]:
simple_lm.intercept_

-7789.150445273459

In [30]:
simple_lm.coef_.item()

166.58375411113175

ESTIMATED AMOUNT

    - Price = -7789.150445273459(simple_lm.intercept_) + 166.58375411113175(simple_lm.coef_.item()) * engine-size